arch/xtensa/core/syscall_helper.c

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright (c) 2022 Intel Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/arch/xtensa/syscall.h>
+
+uintptr_t arch_syscall_invoke6_helper(uintptr_t arg1, uintptr_t arg2,
+				      uintptr_t arg3, uintptr_t arg4,
+				      uintptr_t arg5, uintptr_t arg6,
+				      uintptr_t call_id)
+{
+	register uintptr_t a2 __asm__("%a2") = call_id;
+	register uintptr_t a6 __asm__("%a6") = arg1;
+	register uintptr_t a3 __asm__("%a3") = arg2;
+	register uintptr_t a4 __asm__("%a4") = arg3;
+	register uintptr_t a5 __asm__("%a5") = arg4;
+	register uintptr_t a8 __asm__("%a8") = arg5;
+	register uintptr_t a9 __asm__("%a9") = arg6;
+
+	__asm__ volatile("syscall\n\t"
+			 : "=r" (a2)
+			 : "r" (a2), "r" (a6), "r" (a3), "r" (a4),
+			   "r" (a5), "r" (a8), "r" (a9)
+			 : "memory");
+
+	return a2;
+}
+
+uintptr_t arch_syscall_invoke5_helper(uintptr_t arg1, uintptr_t arg2,
+				      uintptr_t arg3, uintptr_t arg4,
+				      uintptr_t arg5, uintptr_t call_id)
```

Comment:
Regarding these being functions and not inlines... "we couldn't make it work with the compiler" seems like pretty weak justification. :)

If it's a bug, can we get it isolated and reported to Cadence or whoever?  Syscalls really (really) want to be fast, and having to take the extra function call to swap all those registers around (and then do it again because their order doesn't match the syscall ABI!, see comment elsewhere) seems really expensive without good justification.

Reply:
Yep, we are looking it. But if remember correctly we saw this happen even with GCC. Initial implementation was obviously inlining it, but in several places the generate code was not understanding that registers had changed in the inline assembler and was using them as if they hadn't changed. Using the stack would have an impact too because we would need `alloca` to manipulate it.

Reply:
Just to give some context, when I was starting to implement it I was looking Linux and that is the xtensa syscall ABI used there. I got it even before the implementation itself. Just changed to use the window ABI convention.